framework/include/loops/ComputeFVFluxThread.h

```diff
@@ -412,15 +433,138 @@ ComputeFVFluxThread<RangeType>::post()
   _fe_problem.clearActiveMaterialProperties(_tid);
 }
 
+template <typename RangeType>
+void
+ComputeFVFluxThread<RangeType>::emptyDifferenceTest(const std::set<unsigned int> & requested,
+                                                    const std::set<unsigned int> & supplied,
+                                                    std::set<unsigned int> & difference)
+{
+  std::set_difference(requested.begin(),
+                      requested.end(),
+                      supplied.begin(),
+                      supplied.end(),
+                      std::inserter(difference, difference.begin()));
+
+  mooseAssert(
+      difference.empty(),
+      "All of the material properties we depend on should already be supplied/computed. Do your FV "
+      "objects depend on a material property that is computed in a material with coupled FE "
+      "variables? If so, that property needs to be moved to a material without FE coupling.");
+}
+
+template <typename RangeType>
+void
+ComputeFVFluxThread<RangeType>::checkPropDeps(
+    const std::vector<std::shared_ptr<MaterialBase>> & libmesh_dbg_var(mats)) const
+{
+#ifndef NDEBUG
+  std::set<unsigned int> props_diff;
+  std::set<unsigned int> supplied_props;
+  std::set<unsigned int> fv_kernel_requested_props;
+
+  for (auto * kernel : _fv_flux_kernels)
+  {
+    const auto & mp_deps = kernel->getMatPropDependencies();
+    fv_kernel_requested_props.insert(mp_deps.begin(), mp_deps.end());
+  }
+
+  for (std::shared_ptr<MaterialBase> mat : mats)
+  {
+    for (const auto prop_id : mat->getSuppliedPropIDs())
+    {
+      auto pr = supplied_props.insert(prop_id);
+      mooseAssert(pr.second,
+                  "Multiple objects supply property ID "
+                      << pr.second
+                      << ". Do you have materials on neighboring subdomains that define the same "
```

Comment:
I didn't like doing this either. But how do I handle this?
````
[FVFluxKernels]
  [fluid_kernel]
    type = FluidKernel
    rho = rho
    mu = mu
    block = 0
  []
[]

[Materials]
  [fluid_mat]
    type = GenericConstantMaterial
    block = 0
    prop_names = 'rho mu'
    prop_values = '1 1'
  []
  [solid_mat]
    type = GenericConstantMaterial
    prop_names = 'rho'
    prop_values = '1000'
  []
[]
```

Reply:
Moreover, even if I did, how do I tell `fluid_mat` to only compute `mu` on the interface but not `rho`?

Reply:
Why wouldn't you want to use the neighboring subdomain's material there?  If you didn't want to use the neighboring subdomain's material in this case, then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid.  But when they are the same name, I think we should treat them as one continuous material prop field.

Reply:
Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

Reply:
>  What if I have 4 subdomains, and for one material property I want a different constant generic property value on each?

and

>  then rho represents something different in each subdomain and the user should correspondingly name it differently - like rho_fluid and rho_solid

seem inconsistent.

> Why wouldn't you want to use the neighboring subdomain's material there?

Because the `fluid_kernel` is restricted to block 0. The user clearly intends to only use material properties from block 0. However if they are using Dirichlet BCs, they are going to need to ghost value for rho on the neighboring side of the interface between blocks 0 and 1. So I want to make sure that `fluid_mat` gets evaluated on the neighbor side of the interface between block 0 and 1. However, if I am also evaluating a material property `rho` in `solid_mat`, then there is a race condition there.

Reply:
> Wouldn't you still need mu on the interface? - you just need to ghost it instead of using the property from the neighboring subdomain (like you can with rho).

So you're talking about whole new code for material property evaluation that would happen outside of the `Material` class. This is probably what you were thinking when we talked on the phone. That's reasonable. *But* I still think we should make it impossible for the `fluid_kernel` to use the `solid_mat` value for `rho` because that's clearly not what the user intended (because of his/her choice for block restriction). What I've implemented here accomplishes that.

Reply:
Yes but what you implemented here also makes impossible a very common use case.  I'm saying expecting users to supply a different property name for regions with different physics doesn't seem unreasonable.  However, the user might not have written the kernels.  And might unwittingly use a neighbor material property when they expected ghosting.  Hmmm.  Downsides to all approaches here.  I'll need to think on it a bit.

Reply:
> However, the user might not have written the kernels.

The issue is with the suppliers right? So there definitely could be a problem if the user wants to use existing `Materials` that hard-code a property name. My solution for that would be to go update those `Materials` and make them more flexible. 

> Yes but what you implemented here also makes impossible a very common use case

I do agree. I think it's totally reasonable for a user to define 'k' on multiple blocks. It's at the core of the physical concept of a material. I'm defining a thermal conductivity. Thermal conductivity is denoted by 'k'. I should be able to define 'k' for every material in my simulation.

But in the end I'd rather make the user do more work in the input file than make it possible for the simulation to do something wrong/unexpected.

If there's a straightforward implementation that accomplishes both: 1) guaranteed safety and 2) user convenience, then I'm all for it. I haven't thought of it, but maybe you will.

Ping me on slack if you want to talk about it over the phone.